### PR TITLE
Add GCM Benchmarks

### PR DESCRIPTION
**Pull request by thomasgibson**
This PR is co-authored by: @bischtob, @mwarusz, and @akshaysridhar, and supersedes both #1193 and #1205. We have added a suite of experiments for the GCM model. This includes the following:
1. A non-hydrostatic gravity wave,
2. A balanced isothermal flow,
3. A baroclinic wave;
4. An updated Held-Suarez experiment (with tracer-support).

This PR addresses issue #1306, but we still need to sort out how we will test these experiments (similarly for the LES) in order to ensure we don't regress.

# Description

A clear and concise description of the code with usage.

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by jkozdon** [comment]
> This PR addresses issue #1306, but we still need to sort out how we will test these experiments (similarly for the LES) in order to ensure we don't regress.

I would take a look at what @blallen and @christophernhill have been doing for the ocean model with @christophernhill's `StateCheck` functionality. See [here](https://github.com/CliMA/ClimateMachine.jl/blob/1171a998ca49b9e7a3efc92309ecfd4797110428/test/Ocean/HydrostaticBoussinesq/3D_hydrostatic_spindown.jl#L131-L135) for example.

**Comment by thomasgibson** [comment]
> > This PR addresses issue #1306, but we still need to sort out how we will test these experiments (similarly for the LES) in order to ensure we don't regress.
> 
> I would take a look at what @blallen and @christophernhill have been doing for the ocean model with @christophernhill's `StateCheck` functionality. See [here](https://github.com/CliMA/ClimateMachine.jl/blob/1171a998ca49b9e7a3efc92309ecfd4797110428/test/Ocean/HydrostaticBoussinesq/3D_hydrostatic_spindown.jl#L131-L135) for example.

Thanks for the information! I'll take a look.

File: experiments/AtmosGCM/heldsuarez.jl [hunk 1]
```diff
@@ -1,121 +1,137 @@
...
 const param_set = EarthParameterSet()
 
-struct HeldSuarezDataConfig{FT}
-    T_ref::FT
-end
+import CLIMAParameters
+#CLIMAParameters.Planet.planet_radius(::EarthParameterSet) = 6.371e6 / 125.0
 
-function init_heldsuarez!(bl, state, aux, coords, t)
+function init_baroclinic_wave!(bl, state, aux, coords, t)
```
Comment (tapios) [inline experiments/AtmosGCM/heldsuarez.jl:33]: Why initialize HS with this detailed a baroclinic wave?

File: experiments/AtmosGCM/heldsuarez-tracers.jl [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        domain_height,
+        param_set,
+        init_heldsuarez!;
+        model = model,
+    )
+
+    return config
+end
+
+function held_suarez_forcing!(
```
Comment (tapios) [inline experiments/AtmosGCM/heldsuarez-tracers.jl:120]: Can we abstract this out somewhere, so we do not need to duplicate code here and in heldsuarez.jl? I'd prefer for this forcing to be defined and coded in only one place.

File: experiments/AtmosGCM/heldsuarez.jl [hunk 3]
```diff
@@ -173,73 +189,85 @@ function held_suarez_forcing!(
     k_v = k_f * height_factor
 
     # Apply Held-Suarez forcing
-    source.ρu -= k_v * projection_tangential(bl, aux, ρu)
+    source.ρu -= k_v * ρu
```
Comment (tapios) [inline experiments/AtmosGCM/heldsuarez.jl:192]: Why are we back to damping vertical velocities too, rather than just horizontal?

Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez.jl:192]: It looks like they damp all velocity components in Held & Suarez 1994.

File: experiments/AtmosGCM/heldsuarez-tracers.jl (hunk 2 above)
Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez-tracers.jl:120]: I agree. I don't know where we would put this very specific forcing function. Would need to create a separate space.

File: experiments/AtmosGCM/heldsuarez.jl (hunk 1 above)
Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez.jl:33]: That's a mistake. Needs to be reverted.

**Comment by tapios** [comment]
I'm glad to see this coming together. 

The discrete hydrostatic balance issue in #1306 still remains to be addressed, right? 

More generally (not for this PR), we need a cleaner separation of initial conditions/boundary conditions/forcing ("experiment setup") and "model configuration" concerns (LES vs GCM vs single stack, at different resolutions etc.). It would be good to be able to run these different experiments with less duplication of boiler plate code. We'd want a composable structure that allows us to combine different experiment setups with different model configurations. This PR is not the place to do this, but it would be good to get it on people's radar.

File: experiments/AtmosGCM/heldsuarez.jl (hunk 3 above)
Comment (tapios) [reply experiments/AtmosGCM/heldsuarez.jl:192]: No, they don't. The model is hydrostatic, hence `v` is horizontal velocity only.

Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez.jl:192]: Makes sense. I'll revert to original state.

File: experiments/AtmosGCM/heldsuarez-tracers.jl [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+using ClimateMachine
+using ArgParse
+
+s = ArgParseSettings()
+@add_arg_table! s begin
+    "--number-of-tracers"
+    help = "Number of dummy tracers"
+    metavar = "<number>"
+    arg_type = Int
+    default = 0
```
Comment (thomasgibson) [inline experiments/AtmosGCM/heldsuarez-tracers.jl:11]: @bischtob Why do we have two versions of held suarez? by default, no tracers are set in this script. If the only functional difference between this held suarez and the other one is the tracer argument, why not just add this tracer stuff to `heldsuarez.jl`?

Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez-tracers.jl:11]: Works for me!

File: experiments/AtmosGCM/heldsuarez-tracers.jl (hunk 2 above)
Comment (bischtob) [reply experiments/AtmosGCM/heldsuarez-tracers.jl]: Perhaps we just combine the two drivers into one.

**Comment by bischtob** [comment]
> I'm glad to see this coming together.
> 
Thanks!

> The discrete hydrostatic balance issue in #1306 still remains to be addressed, right?
> 
Yes, @mwarusz is putting this together as he is putting together the Roe fluxes in #1339.

File: experiments/AtmosGCM/heldsuarez-tracers.jl (hunk 2 above)
Comment (thomasgibson) [reply experiments/AtmosGCM/heldsuarez-tracers.jl]: Done!

**Comment by bischtob** [comment]
@simonbyrne, docs breaking?

**Comment by jakebolewski** [comment]
fixed by #1362

**Comment by thomasgibson** [comment]
bors try

**Comment by bischtob** [comment]
bors try

**Comment by blallen** [comment]
> ## try
> Build failed:
> 
> * [docs-build](https://github.com/CliMA/ClimateMachine.jl/runs/894368212)

I think you need to rebase on top of master for CI to work properly

**Comment by bischtob** [comment]
bors try

**Comment by simonbyrne** [comment]
LGTM. 

The only question is how we want to test this: these aren't currently hooked up to the Slurm CI pipeline.

@thomasgibson @bischtob what did you have in mind?

**Comment by simonbyrne** [comment]
Can you coordinate the @jakebolewski to figure out what you want to do with this?

**Comment by jakebolewski** [comment]
Is there a reduced problem size that we could run just to check that it doesn't trivially break (with some code changes) and run that with Slurm-CI? 

For real runs that we want to interpret I think I would just need the driver files / slurm batch scripts you are using to generate the output.

**Comment by bischtob** [comment]
> Is there a reduced problem size that we could run just to check that it doesn't trivially break (with some code / api changes) and run that with Slurm-CI?
> 
> For real runs that we want to interpret I think I would just need the driver files / slurm batch scripts you are using to generate the output.

@jiahe23 & @LenkaNovak are the best to decide :-). They are running the GCM.

**Comment by tapios** [comment]
> > Is there a reduced problem size that we could run just to check that it doesn't trivially break (with some code / api changes) and run that with Slurm-CI?
> > For real runs that we want to interpret I think I would just need the driver files / slurm batch scripts you are using to generate the output.
> 
> @jiahe23 & @LenkaNovak are the best to decide :-). They are running the GCM.

Just run ~10 time steps?

**Comment by jiahe23** [comment]
> > > Is there a reduced problem size that we could run just to check that it doesn't trivially break (with some code / api changes) and run that with Slurm-CI?
> > > For real runs that we want to interpret I think I would just need the driver files / slurm batch scripts you are using to generate the output.
> > 
> > 
> > @jiahe23 & @LenkaNovak are the best to decide :-). They are running the GCM.
> 
> Just run ~10 time steps?

@jakebolewski 

10 steps sounds reasonable for a Slurm-CI test

An example of bash script we (mainly Lenka and myself) use a lot in visualize GCM output is at the [wiki page](https://github.com/CliMA/ClimateMachine.jl/wiki/Bash-Run-Scripts) which only output zonal mean zonal wind for now. 

Define the diagnostics/visualization for CI is on-going here #1363

File: src/Common/Orientations/Orientations.jl [hunk 5]
```diff
@@ -141,6 +143,57 @@ latitude(orientation::SphericalOrientation, aux::Vars) =
...
+        clat * vec[2] + slat * vec[3],
+    )
+
+    return u
+end
+
+"""
+    cart_to_sphr_vec(orientation::SphericalOrientation, state::Vars, aux::Vars)
+
+Projects a vector defined based on unit vectors in cartesian coordinates to a spherical unit vectors.
```
Comment (charleskawczynski) [inline src/Common/Orientations/Orientations.jl:175]:
```suggestion
Projects a vector defined based on unit vectors in cartesian coordinates to a spherical unit vector.
```

File: src/Common/Orientations/Orientations.jl [hunk 6]
```diff
@@ -141,6 +143,57 @@ latitude(orientation::SphericalOrientation, aux::Vars) =
 longitude(orientation::SphericalOrientation, aux::Vars) =
     @inbounds atan(aux.coord[2], aux.coord[1])
 
+"""
+    sphr_to_cart_vec(orientation::SphericalOrientation, state::Vars, aux::Vars)
+
+Projects a vector defined based on unit vectors in spherical coordinates to cartesian unit vectors.
```
Comment (charleskawczynski) [inline src/Common/Orientations/Orientations.jl:149]:
```suggestion
Projects a vector defined based on unit vectors in spherical coordinates to cartesian unit vector.
```

**Pull request merged**